### PR TITLE
feat: eth: optimize eth block loading + eth_feeHistory

### PR DESCRIPTION
**Pull request by Stebalien**
## Related Issues

The current implementations of eth_getBlock* and eth_feeHistory have proven computationally expensive due to their usage of full tipset computations (`StateCompute`). This PR replaces the expensive logic with receipt lookups.

## Proposed Changes

For eth_getBlock* and eth_feeHistory, instead of replaying the tipset:

1. Lookup the tipset receipts.
2. Load the receipts.
3. Calculate everything from there (we have the gas used, etc.)

This also removes support for the `pending` block parameter as it's impossible to handle it without forcing state computes, which are prohibitively expensive. The Eth community is evaluating a proposal to also remove it in their next fork, given its limited applications due to MEV and other reasons.

## Additional Info

We've gone down from 10 seconds to various minutes (depending on the method and amount of data request) to almost instantaneous responses.

### After

```
⟩ time curl --silent --url http://localhost:1234/rpc/v1 \
      -X POST \
      -H "Content-Type: application/json" \
      -d '{"jsonrpc":"2.0","method":"eth_feeHistory","params":[ "0x08", "latest", [10, 50, 90] ],"id":1}'
{"jsonrpc":"2.0","result":{"oldestBlock":"0x28d39e","baseFeePerGas":["0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b","0x56ceb9b"],"gasUsedRatio":[1.1866695961,2.0196181532,2.5688713134,1.6794295302,1.7004244153,1.2444834388,1.736932316,2.2669179384],"reward":[["0x2de55df","0x2e29630","0x2e8e694"],["0x2d92b3b","0x2e665f8","0x5a874a8"],["0x294d17d","0x2e5547e","0x2eb2dbb"],["0x11c1e0d","0x2e35135","0x38a8b8f"],["0x11b8225","0x2e4b13e","0x39fb27c"],["0x2e18bb6","0x2e7c695","0x5d51c6a"],["0x2e0836d","0x2e70add","0x2ec2445"],["0x2e18763","0x2e429c9","0x2ebcc8a"]]},"id":1}

________________________________________________________
Executed in  345.14 millis    fish           external
   usr time    3.56 millis    0.12 millis    3.44 millis
   sys time    7.72 millis    3.36 millis    4.36 millis

~/W/pl/lotus · (steb/optimize-eth-block±)
⟩ time curl --silent --url http://localhost:1234/rpc/v1 \
      -X POST \
      -H "Content-Type: application/json" \
      -d '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":[ "latest", false ],"id":1}'
...

________________________________________________________
Executed in  366.47 millis    fish           external
   usr time    4.64 millis    0.12 millis    4.52 millis
   sys time   11.66 millis    4.58 millis    7.08 millis
```

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by Stebalien** [comment]
Blocked on #10445.

File: itests/eth_block_hash_test.go [hunk 1]
```diff
@@ -49,9 +49,6 @@ func TestEthBlockHashesCorrect_MultiBlockTipset(t *testing.T) {
 	// let the chain run a little bit longer to minimise the chance of reorgs
 	n2.WaitTillChain(ctx, kit.HeightAtLeast(head.Height()+50))
 
-	head, err = n2.ChainHead(context.Background())
-	require.NoError(t, err)
```
Comment (raulk) [inline itests/eth_block_hash_test.go:53]: Any particular reason to remove this?

**Review by raulk — APPROVED** [review]
This looks good, about to test it.

File: itests/eth_block_hash_test.go (hunk 1 above)
Comment (Stebalien) [reply itests/eth_block_hash_test.go:53]: It defeats the point of waiting 50 blocks for the chain to stabilize.

Comment (raulk) [reply itests/eth_block_hash_test.go:53]: Not sure I follow the thinking? This is just a sanity check. Might be worthless but what point does it defeat exactly?

Comment (Stebalien) [reply itests/eth_block_hash_test.go:53]: Unless we wait a bit after getting the head, there's no guarantee that `EthGetBlockByNumber` and `EthGetBlockByHash` will return the same value at the same height due to reorgs.

I _thought_ that that was the entire point of waiting an extra 50 blocks. If that wasn't, the comment:

> let the chain run a little bit longer to minimise the chance of reorgs

Is wrong.

Comment (raulk) [reply itests/eth_block_hash_test.go:53]: I see, all cleared up. `head` is being used below to traverse the chain up until that point. We do not need to traverse the extra 50 tipsets, those are there for stability. This is fine.

**Review by raulk — APPROVED** [review]
I migrated `eth_feeHistory` to reading receipts, and dropped support for the `pending` block parameter. Fleshed out the description and added numbers. This is ready to go.

**Comment by raulk** [comment]
@Stebalien can you review my changes?

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -689,49 +689,43 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
...
 		if err != nil {
-			return ethtypes.EthFeeHistory{}, xerrors.Errorf("cannot lookup the status of tipset: %v: %w", ts, err)
+			return ethtypes.EthFeeHistory{}, xerrors.Errorf("failed to retrieve messages and receipts for height %d: %w", ts.Height(), err)
 		}
 
 		txGasRewards := gasRewardSorter{}
-		for _, msg := range compOutput.Trace {
-			if msg.Msg.From == builtintypes.SystemActorAddr {
+		for i, msg := range msgs {
+			if msg.VMMessage().From == builtintypes.SystemActorAddr {
```
Comment (Stebalien) [inline node/impl/full/eth.go:713]: This isn't necessary anymore. The messages in the tipset will never be from the system.

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -689,49 +689,43 @@ func (a *EthModule) EthFeeHistory(ctx context.Context, p jsonrpc.RawParams) (eth
...
+		baseFeeArray      = []ethtypes.EthBigInt{ethtypes.EthBigInt(basefee)}
+		rewardsArray      = make([][]ethtypes.EthBigInt, 0)
+		gasUsedRatioArray = []float64{}
+		blocksIncluded    int
+	)
 
-	blocksIncluded := 0
 	for blocksIncluded < int(params.BlkCount) && ts.Height() > 0 {
-		compOutput, err := a.StateCompute(ctx, ts.Height(), nil, ts.Key())
+		msgs, rcpts, err := messagesAndReceipts(ctx, ts, a.Chain, a.StateAPI)
```
Comment (Stebalien) [inline node/impl/full/eth.go:706]: This works but it isn't great. Ideally we'd just walk the chain backwards as described in #10288.

**Review by Stebalien — COMMENTED** [review]
I can't approve but the new changes look correct.

**Comment by Stebalien** [comment]
It looks like we need to fix the conformance tests to drop the "pending" one.

File: node/impl/full/eth.go (hunk 3 above)
Comment (raulk) [reply node/impl/full/eth.go:706]: Aren't the mechanics the same?

**Comment by icodeface** [comment]
How to get correct nonce with pending txs after "pending" param is dropped ?

**Comment by go-lifei** [comment]
> How to get correct nonce with pending txs after "pending" param is dropped ?

        let nonce = await ethers.provider.getTransactionCount(
            addr,
            'latest'
        )

**Comment by icodeface** [comment]
> getTransactionCount

This doesn't take into transactions in the mempool, which could cause nonce collisions.

**Pull request merged**